fix: prevent hard javascript errors when async JS-plugins cannot be fetched

tobiasberge opened the pull request:
closes #11592
<!--
Thank you for contributing to Shopware! Please fill out this description template to help us process your pull request.

Please make sure to fulfil our contribution guidelines (https://developer.shopware.com/docs/resources/guidelines/code/contribution?category=shopware-platform-dev-en/contribution).

Do your changes need to be documented?
Please create a second pull request at https://github.com/shopware/docs
-->

### 1. Why is this change necessary?


### 2. What does this change do, exactly?


### 3. Describe each step to reproduce the issue or behaviour.


### 4. Please link to the relevant issues (if any).
<!-- Examples:
- closes #123  - closes the issue #123 when the PR is merged
- relates #123 - relates to the issue #123

If the issue exists only in Jira, include a link to the Jira issue.
- Jira issue: https://shopware.atlassian.net/browse/NEXT-123
-->

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [ ] I have created a [changelog file](https://github.com/shopware/shopware/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [ ] I have read the contribution requirements and fulfilled them

tobiasberge committed:
fix: prevent hard javascript errors when async JS-plugins cannot be fetched

raphael-homann commented:
Here you handle network error (or missing file).

There is another case that can happend : if an exception is thrown (or any error) in the plugin. In this case, `fetchedPluginClasses[index]` exists, but is instance of `Error`

It is possible to protect bu using : 
```
            if (!fetchedPluginClasses[index]) {
                console.error(`The plugin "${plugin.name}" could not be fetched.`);
                return;
            } else if (fetchedPluginClass instanceof Error) {
                console.error(`Plugin "${plugin.pluginName}" could not be fetched: ${fetchedPluginClass.message}`);
                return;
            }
```

it is possible to trigger this behavior with this hack : 
```
        queue.push({
            pluginClassPromise: () => new Promise((resolve) => {
                console.log('Hello');
                missintFunctionWillFail();
            })
        });
```

tobiasberge committed:
use Promise.allSettled for more graceful error-handling

tobiasberge commented:
@raphael-homann Thanks for your suggestion. I did quite some testing and added your case as well.

However, for me the first "could not be fetched" case (`if (!fetchedPluginClasses[index])`) does not work as expected and I basically always land in the error case. Because a file that is not found also triggers a rejected promise including an error. And since we now .catch and pipe through all the inner errors in Promise.all, `fetchedPluginClasses[index]` contains the error and is not "undefined". It is only "undefined" when the Promise is resolved with "nothing". But I am not sure if this can even be the case in the real world without faking this with some mock data.

I did some more research and think that [Promise.allSettled](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) can be a good approach for us. Unlike `Promise.all` it does not "break" whenever one of the promises rejects. Instead it runs through first and collects all fulfilled and rejected promises, that we can then handle later in the `queue.forEach`. I updated the PR including some comments. Does this also handle your issues?

<img width="1156" height="393" alt="image" src="https://github.com/user-attachments/assets/797aa478-e24b-4bf7-bb6e-09cd573ffb85" />

tobiasberge committed:
add a test plugin to test throws

raphael-homann commented:
Nice. It looks like you covered a lot of different cases ;) 
the [Promise.allSettled](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) could be a good thing too, I don't know it enough... we just have to be sure every exception is covered, not only rejected promises.

in your latest version, I think we have the risk that `fetchedPluginClasses[index]` is undefined, so `fetchedItem.status` will fail. This will lead to the initial problem we are trying to solve here...

I think it would be safer to check if `fetchedItem` exists (or just if `fetchedPluginClasses` has `index`

```
const fetchedItem = fetchedPluginClasses[index];

            // If the promise from allSettled is rejected, we can be sure the plugin could not be fetched.
            if (fetchedItem.status === 'rejected') {
                console.warn(`The plugin "${plugin.pluginName}" could not be fetched. Reason: `, fetchedItem.reason);
                return;
            }
```

tobiasberge committed:
Merge branch 'trunk' into async-plugin-imports
